src/in-depth/machine-communication-stdin.rs

```diff
@@ -0,0 +1,34 @@
+use clap::Parser;
+use std::io::stdin;
+
+/// Count the number of lines in a file
+#[derive(Parser)]
+#[command(arg_required_else_help = true)]
+struct Cli {
+    /// The path to the file to read, use - to read from stdin
+    file: String,
+}
+
+fn main() {
+    let args = Cli::parse();
+
+    let mut word_count = 0;
+    let mut file = args.file;
+
+    if file == "-" {
+        file = String::from("stdin");
+        for line in stdin().lines() {
+            let line = line.unwrap();
+            if !line.trim().is_empty() {
+                word_count += line.split(' ').count();
+            }
+        }
+    } else {
+        let content = std::fs::read_to_string(&file).unwrap();
+        for line in content.lines() {
+            word_count += line.split(' ').count();
+        }
+    }
```

Comment:
We should probably be showing how to share logic between stdin/file as that is what users will likely want to do

Reply:
I'm struggling with how to go about sharing this logic since I'm new to Rust and the `Lines` returned from reading the file and reading from stdin are different types. Are you able to provide some guidance on what that code could look like?

Reply:
I had some success with figuring out how to do this in https://github.com/rust-cli/book/pull/180/commits/90da0f56be17ca2e72e175ba5332bff7b9d467e9 by using `BufReader`. I'm not sure the code is great, but it's got the logic for counting shared.